### PR TITLE
#170 If the user is not logged in we get NPE

### LINKED ISSUE #170: If user is not logged in we get NPE
[Link to comment](https://github.com/intive/Picover/pull/157#discussion_r1313816378)

### PR DESCRIPTION
**Pull request by kmaslowiec**

### CODE REVIEW

**Comment by kmaslowiec** [comment]
I've updated the name of the task to reflect the issue that was addressed. Editing the name of the branch in Github deletes the PR.

**Comment by BorysChajdas** [comment]
@kmaslowiec (for the future) In such cases it's better to update pull request (even change the title) than to close and open new one because we are losing some of the history

File: app/src/main/kotlin/com/intive/picover/auth/repository/AuthRepository.kt [hunk 1]
```diff
@@ -80,6 +81,6 @@ class AuthRepository @Inject constructor(
 		return userProfile()
 	}
 
-	private fun requireUser() =
+	private fun requireUser(): FirebaseUser =
```
Comment (BorysChajdas) [inline app/src/main/kotlin/com/intive/picover/auth/repository/AuthRepository.kt]: Is an explicit type here needed? Why was it added `FirebaseUser`?

**Review by BorysChajdas — APPROVED** [review]

File: app/src/main/kotlin/com/intive/picover/auth/repository/AuthRepository.kt (hunk 1 above)
Comment (kmaslowiec) [reply app/src/main/kotlin/com/intive/picover/auth/repository/AuthRepository.kt]: I think it was added when I converted to the block body when I worked on the code. I overlooked it when I returned it to the previous version. Thank you.

**Comment by kmaslowiec** [comment]
> @kmaslowiec (for the future) In such cases it's better to update pull request (even change the title) than to close and open new one because we are losing some of the history

@BorysChajdas answered in closed PR

**Review by michal-kucznerowicz — APPROVED** [review]

File: app/src/test/kotlin/com/intive/picover/auth/repository/AuthRepositoryTest.kt [hunk 2]
```diff
@@ -129,17 +129,12 @@ class AuthRepositoryTest : ShouldSpec(
 			tested.updateUserAvatar(userPhoto) shouldBe profile
 		}
 
-		should("return Profile WHEN userProfile called") {
-			coEvery { referenceToUserAvatar.downloadUrl.await() } returns userPhoto
-
-			tested.userProfile() shouldBe profile
-		}
```
Comment (MarekMacko) [inline app/src/test/kotlin/com/intive/picover/auth/repository/AuthRepositoryTest.kt]: Why this has been removed?

File: app/src/main/kotlin/com/intive/picover/auth/repository/AuthRepository.kt [hunk 3]
```diff
@@ -20,7 +20,7 @@ class AuthRepository @Inject constructor(
 	private val bitmapConverter: BitmapConverter,
 ) {
 
-	private val userAvatarReference = storageReference.child("user/${requireUser().uid}")
+	private val userAvatarReference by lazy { storageReference.child("user/${requireUser().uid}") }
```
Comment (MarekMacko) [inline app/src/main/kotlin/com/intive/picover/auth/repository/AuthRepository.kt:23]: If someone would change it, intentionally or by mistake, to ordinary property (previous form), then all tests will still be green.   
The goal of unit testing is to detect such cases, and prevent introducing the same bug once again.
Conclusion: add unit test

**Review by MarekMacko — CHANGES_REQUESTED** [review]

File: app/src/test/kotlin/com/intive/picover/auth/repository/AuthRepositoryTest.kt (hunk 2 above)
Comment (kmaslowiec) [reply app/src/test/kotlin/com/intive/picover/auth/repository/AuthRepositoryTest.kt]: Honestly, deleted by mistake. Thank you

File: app/src/main/kotlin/com/intive/picover/auth/repository/AuthRepository.kt (hunk 3 above)
Comment (kmaslowiec) [reply app/src/main/kotlin/com/intive/picover/auth/repository/AuthRepository.kt:23]: Is going to be addressed.

File: app/src/test/kotlin/com/intive/picover/auth/repository/AuthRepositoryTest.kt (hunk 2 above)
Comment (MarekMacko) [reply app/src/test/kotlin/com/intive/picover/auth/repository/AuthRepositoryTest.kt]: Then please convert PR to draft if there is still ongoing work on your side

**Review by marcinWisnia — APPROVED** [review]

File: app/src/main/kotlin/com/intive/picover/auth/repository/AuthRepository.kt (hunk 3 above)
Comment (kmaslowiec) [reply app/src/main/kotlin/com/intive/picover/auth/repository/AuthRepository.kt:23]: @MarekMacko Test added

File: app/src/test/kotlin/com/intive/picover/auth/repository/AuthRepositoryTest.kt [hunk 4]
```diff
@@ -151,5 +152,9 @@ class AuthRepositoryTest : ShouldSpec(
 			result shouldBeEqual Profile(userPhoto, "Marian Kowalski", userEmail)
 			slot.captured.displayName!! shouldBeEqual "Marian Kowalski"
 		}
+
+		should("not set userAvatarReference WHEN class is created") {
+			verify(inverse = true) { storageReference.child("user/$userUid") }
```
Comment (MarekMacko) [inline app/src/test/kotlin/com/intive/picover/auth/repository/AuthRepositoryTest.kt]: I would rather simplify it and use regular `verify { mock wasNot Called }` and assert that any call to `storageReference` was performed, regardless of used function and passed arguments. 
If someone would change parameter of `child` call in `AuthRepository` and change it to ordinary property, then test would not detect it.

**Review by MarekMacko — CHANGES_REQUESTED** [review]

File: app/src/test/kotlin/com/intive/picover/auth/repository/AuthRepositoryTest.kt [hunk 5]
```diff
@@ -151,5 +153,9 @@ class AuthRepositoryTest : ShouldSpec(
 			result shouldBeEqual Profile(userPhoto, "Marian Kowalski", userEmail)
 			slot.captured.displayName!! shouldBeEqual "Marian Kowalski"
 		}
+
+		should("not set storageReference WHEN class is created") {
```
Comment (MarekMacko) [inline app/src/test/kotlin/com/intive/picover/auth/repository/AuthRepositoryTest.kt]: part of the name "not set" seems to be incorrect, we're not checking any "set" here, it's rather "call"

File: app/src/test/kotlin/com/intive/picover/auth/repository/AuthRepositoryTest.kt (hunk 4 above)
Comment (kmaslowiec) [reply app/src/test/kotlin/com/intive/picover/auth/repository/AuthRepositoryTest.kt]: @MarekMacko Updated

File: app/src/test/kotlin/com/intive/picover/auth/repository/AuthRepositoryTest.kt (hunk 5 above)
Comment (kmaslowiec) [reply app/src/test/kotlin/com/intive/picover/auth/repository/AuthRepositoryTest.kt]: @MarekMacko Updated

Comment (MarekMacko) [reply app/src/test/kotlin/com/intive/picover/auth/repository/AuthRepositoryTest.kt]: Nice, thank you ;)

File: app/src/test/kotlin/com/intive/picover/auth/repository/AuthRepositoryTest.kt (hunk 4 above)
Comment (MarekMacko) [reply app/src/test/kotlin/com/intive/picover/auth/repository/AuthRepositoryTest.kt]: Nice, thank you ;)

**Review by MarekMacko — APPROVED** [review]
Ready to go!

**Pull request merged**